File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_cpp/code_output/IPAddress.cpp
#include <string>
#include <vector>
#include <sstream>
#include <bitset>
#include <stdexcept>

class IPAddress {
private:
    std::string ipAddress;

    static std::vector<std::string> split(const std::string& s, char delimiter) {
        std::vector<std::string> tokens;
        std::string token;
        std::istringstream tokenStream(s);
        while (std::getline(tokenStream, token, delimiter)) {
            tokens.push_back(token);
        }
        return tokens;
    }

public:
    IPAddress(const std::string& ipAddress) : ipAddress(ipAddress) {}

    bool isValid() const {
        std::vector<std::string> octets = split(ipAddress, '.');
        if (octets.size() != 4) {
            return false;
        }
        for (const std::string& octet : octets) {
            try {
                int num = std::stoi(octet);
                if (num < 0 || num > 255) {
                    return false;
                }
            } catch (const std::invalid_argument&) {
                return false;
            } catch (const std::out_of_range&) {
                return false;
            }
        }
        return true;
    }

    std::vector<std::string> getOctets() const {
        std::vector<std::string> octetsList;
        if (isValid()) {
            octetsList = split(ipAddress, '.');
        }
        return octetsList;
    }

    std::string getBinary() const {
        if (!isValid()) {
            return "";
        }
        std::string binaryString;
        std::vector<std::string> octets = split(ipAddress, '.');
        for (const std::string& octet : octets) {
            int num = std::stoi(octet);
            std::string bin = std::bitset<8>(num).to_string();
            binaryString += bin + ".";
        }
        if (!binaryString.empty()) {
            binaryString.pop_back(); // remove trailing '.'
        }
        return binaryString;
    }
};

File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/cpp_to_py/IPAddress.py
class IPAddress:
    def __init__(self, ip_address: str):
        self.ip_address = ip_address

    def is_valid(self) -> bool:
        octets = self.ip_address.split('.')
        if len(octets) != 4:
            return False
        for octet in octets:
            if not self.is_valid_octet(octet):
                return False
        return True

    def get_octets(self) -> list:
        if self.is_valid():
            return self.ip_address.split('.')
        return []

    def get_binary(self) -> str:
        if self.is_valid():
            octets = self.get_octets()
            binary_parts = []
            for octet in octets:
                num = int(octet)
                binary_parts.append(bin(num)[2:].zfill(8))
            return '.'.join(binary_parts)
        return ''

    def is_valid_octet(self, octet: str) -> bool:
        if octet == "" or len(octet) > 3:
            return False
        if not octet.isdigit():
            return False
        num = int(octet)
        return 0 <= num <= 255